### PR TITLE
Basic testing scaffolding and unit tests for wasm_utils serialization

### PR DESCRIPTION
**Pull request by ddd-mtl**
Added integration testing scaffolding for `wasm_utils` serialization.
All 'unit tests' are done in an integration test that makes use of the Holochain Ribosome.
Each test calls an unique function in wasm.

Fixed potential serious bug in serialization where Out Of Memory was checked **after** writing in memory.

##### API Change @Connoropolous 
Changed `SinglePageStack::from_encoded()` to `SinglePageStack::from_encoded_allocation()` which now returns a `Result` instead of doing an `expect()`.

**RENAMED ALL SERIALIZATION / DESERIALIZATION FUNCTIONS**
serialization -> store_as_json
deserialization ->load_json
etc.

Started doing store_string and load_string but was have problems with it so will finish that in new PR.

Created an assert macro to use in zome code: `zome_assert!`

### CODE REVIEW

File: wasm_utils/src/memory_serialization.rs [hunk 1]
```diff
@@ -52,8 +52,11 @@ pub fn serialize<T: Serialize>(
 ) -> Result<SinglePageAllocation, RibosomeErrorCode> {
     let json_bytes = serde_json::to_vec(&internal).unwrap();
     let json_bytes_len = json_bytes.len();
-    assert!(json_bytes_len < <u16>::max_value() as usize);
-
+    if json_bytes_len > <u16>::max_value() as usize
+        || json_bytes_len as u32 + stack.top() as u32 > <u16>::max_value() as u32
```
Comment (Connoropolous) [inline wasm_utils/src/memory_serialization.rs]: This is the 'potential bug' fix, I think? good :)

File: wasm_utils/wasm-test/integration-test/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "integration-test"
+version = "0.1.0"
+authors = [""]
+
+[lib]
+crate-type = ["cdylib"]
+
+[profile.release]
+panic = "abort"
```
Comment (Connoropolous) [inline wasm_utils/wasm-test/integration-test/Cargo.toml]: what's panic = "abort" do?

File: wasm_utils/src/memory_serialization.rs (hunk 1 above)
Comment (ddd-mtl) [reply wasm_utils/src/memory_serialization.rs]: yes!

File: wasm_utils/wasm-test/integration-test/Cargo.toml (hunk 2 above)
Comment (ddd-mtl) [reply wasm_utils/wasm-test/integration-test/Cargo.toml]: I don't know. It's the default config we have been using.
I'm actually planning on looking into panic! today.

File: wasm_utils/wasm-test/integration-test/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+lto = true
+
+[workspace]
+members = []
+
+[dependencies]
+serde="1"
+serde_derive="1"
+serde_json="1"
```
[the remark this answers is not in the document]
Comment (ddd-mtl) [reply wasm_utils/wasm-test/integration-test/Cargo.toml]: Fixed

File: wasm_utils/src/memory_serialization.rs [hunk 4]
```diff
@@ -1,103 +1,115 @@
...
-) -> Result<T, String> {
+/// TODO #486 - load and store string from wasm memory
+//pub fn load_string(encoded_allocation: u32) -> Result<String, String> {
+//    let maybe_allocation = decode_encoded_allocation(encoded_allocation);
+//    match maybe_allocation {
+//        Err(return_code) => Err(return_code.to_string()),
+//        Ok(allocation) => Ok(load_string_from_allocation(&allocation)),
+//    }
+//}
+///// Write a string in wasm memory
```
[the remark this answers is not in the document]
Comment (ddd-mtl) [reply wasm_utils/src/memory_serialization.rs]: Not this one. Its commented-out comments :)

File: wasm_utils/src/memory_serialization.rs [hunk 5]
```diff
@@ -1,103 +1,115 @@
...
+) -> Result<SinglePageAllocation, RibosomeErrorCode> {
+    if len as u32 + stack.top() as u32 > U16_MAX {
+        return Err(RibosomeErrorCode::OutOfMemory);
+    }
+    let ptr = stack.allocate(len) as *mut c_char;
+    let ptr_safe = unsafe { slice::from_raw_parts_mut(ptr, len as usize) };
+    for (i, byte) in bytes.iter().enumerate() {
+        ptr_safe[i] = *byte as i8;
     }
+    //return Err(RibosomeErrorCode::OutOfMemory);
```
[the remark this answers is not in the document]
Comment (ddd-mtl) [reply wasm_utils/src/memory_serialization.rs]: Yep!
Its my desperate way to debug things in wasm 😭

**Pull request merged**